pkg/apiserver/profiling/profile.go

```diff
@@ -16,21 +16,43 @@ package profiling
 import (
 	"context"
 	"fmt"
+	"time"
+
+	"github.com/goccy/go-graphviz"
 
 	"github.com/pingcap/tidb-dashboard/pkg/apiserver/model"
+	"github.com/pingcap/tidb-dashboard/pkg/apiserver/profiling/fetcher"
 )
 
-func profileAndWriteSVG(ctx context.Context, fts *fetchers, target *model.RequestTargetNode, fileNameWithoutExt string, profileDurationSecs uint) (string, error) {
+func profileAndWriteSVG(ctx context.Context, cm *fetcher.ClientMap, target *model.RequestTargetNode, fileNameWithoutExt string, profileDurationSecs uint) (string, error) {
+	c, err := cm.Get(target.Kind)
+	if err != nil {
+		return "", err
+	}
+
+	var p *profiler
+
 	switch target.Kind {
-	case model.NodeKindTiKV:
-		return fetchFlameGraphSVG(&flameGraphOptions{duration: profileDurationSecs, fileNameWithoutExt: fileNameWithoutExt, target: target, fetcher: &fts.tikv})
-	case model.NodeKindTiFlash:
-		return fetchFlameGraphSVG(&flameGraphOptions{duration: profileDurationSecs, fileNameWithoutExt: fileNameWithoutExt, target: target, fetcher: &fts.tiflash})
-	case model.NodeKindTiDB:
-		return fetchPprofSVG(&pprofOptions{duration: profileDurationSecs, fileNameWithoutExt: fileNameWithoutExt, target: target, fetcher: &fts.tidb})
-	case model.NodeKindPD:
-		return fetchPprofSVG(&pprofOptions{duration: profileDurationSecs, fileNameWithoutExt: fileNameWithoutExt, target: target, fetcher: &fts.pd})
+	case model.NodeKindTiKV, model.NodeKindTiFlash:
```

Comment:
Seems that your separation of `xxxClient` and `target` is not effective: consider the scenario that you want to **add a new fetch target**. This process will be hard and will be very easy to make mistakes. For example, it is easy to miss updating something.. Compilers just won't warn you about these mistakes. You can only discover these mistakes at runtime. This usually indicate an imperfect design.